packages/preferences/src/browser/abstract-resource-preference-provider.ts

```diff
@@ -157,18 +157,18 @@ export abstract class AbstractResourcePreferenceProvider extends PreferenceProvi
                 this.fileExists = false;
                 return { value: '' };
             });
-        this.readPreferencesFromContent(content.value);
+        await this.readPreferencesFromContent(content.value);
```

Comment:
There used to be stern warnings in this file that most of its work had to be done synchronously. I believe I've eliminated most of the reasons that used to be true - and written tests to ensure everything is alright in any case (I hope) - but adding so many `async-awaits` still makes me a little nervous. What is the problem that the new `awaits` aim to solve?

Reply:
Yes, I know about that warning, and I was quite hesitant adding all the `awaits` here. However, it didn't seem to break anything and fixed my usecase:

https://github.com/eclipse-theia/theia/blob/68bb5157fc25c946a34f65833758b2f96ba6f744/packages/core/src/browser/request/browser-request-service.ts#L68-L69

In there, we wait for the preference service to be ready before sending out our first http-request. The request might fail otherwise, if the backend's `http.proxy` setting hasn't been configured yet. On Theia `master` code, this series of events would unfold:

1. `await preferenceService.ready` (this starts the async user preference reading process __without awaiting it__)
2. Sending the http-request __without having configured proxy settings__ (missing user preferences).
3. `preferenceService.onPreferenceChanged` is triggered, since the async user preference reading process has finished. This then triggers changes to the proxy settings, which is too late.

Reply:
In general, I'm in favor of the new code, especially if it doesn't break anything - and I've added tests of various parts so that if they don't break, the system _should_ be working as expected. But there's one point I'm not sure of:

> 1. `await preferenceService.ready` (this starts the async user preference reading process __without awaiting it__)

There are two `async` parts to reading preferences: one is reading the actual content, and the other is waiting for a debounced event to fire. In the current code, the `AbstractResourcePreferenceProvider` _does_ await the first part before declaring itself ready:

https://github.com/eclipse-theia/theia/blob/f89e8265ef1c59427bc8160ffcb3b2e5f2fe7d15/packages/preferences/src/browser/abstract-resource-preference-provider.ts#L60-L61

https://github.com/eclipse-theia/theia/blob/f89e8265ef1c59427bc8160ffcb3b2e5f2fe7d15/packages/preferences/src/browser/abstract-resource-preference-provider.ts#L149-L159

The last line of the second snippet there, the call to `readPreferencesFromContent` is not awaited (and isn't currently `async`) so `ready` _is_ resolved before the first event emission. That's deliberate, if undesirable, because there are a number of cases even in the framework where the emission of initial values was used to initialize values in other services. But, even though the event emission is not `await`ed, it should be possible to access accurate values as soon as the `ready` promise resolves: the internal state of the provider is already up to date before the event is emitted.

https://github.com/eclipse-theia/theia/blob/f89e8265ef1c59427bc8160ffcb3b2e5f2fe7d15/packages/preferences/src/browser/abstract-resource-preference-provider.spec.ts#L80-L97

Given that, for your case, I think it should be sufficient to `await` the `ready` promise and then access the values, rather than adding additional `await`s for the events. Or have you observed that the configured values are not available when the `ready` promise resolves?

Reply:
I see, that makes sense. I reverted the preference changes and use the `ready` promise to retrieve the preference values instead of waiting for the `onPreferenceChanged` event 👍